slickml/optimization.py

```diff
@@ -335,3 +339,158 @@ def plot_optimization_results(self):
+    optimization over search spaces, which may include real-valued, discrete,
+    and conditional dimensions to train a XGBoost model.Main reference is
+    HyperOpt GitHub: (https://github.com/hyperopt/hyperopt)
+    Parameters
+    ----------
+    num_boost_round: int, optional (default=200)
+        Number of boosting round at each fold of xgboost.cv()
+    metrics: str or tuple[str], optional (default=("auc"))
+        Metric used for evaluation at cross-validation
+        using xgboost.cv(). Please note that this is different
+        than eval_metric that needs to be passed to params dict.
+        Possible values are "auc", "aucpr", "error", "logloss"
+    n_splits: int, optional (default=4)
+        Number of folds for cross-validation
+    shuffle: bool, optional (default=True)
+        Flag to shuffle data to have the ability of building
+    early_stopping_rounds: int, optional (default=20)
+        The criterion to early abort the xgboost.cv() phase
+        if the test metric is not improved
+        early_stopping_rounds: int, optional (default=20)
+    verbost: str, optional (default=False)
+        Print evaluation results from model
+
+    Attributes
+    ----------
+    process: class method
+        Performs hyperparameter tunning on input dataset
+    xgb_cv: class method
+        Optimization function for XGBoost utilizing cross-validation
+    """
+
+    def __init__(
+        self,
+        params=None,
+        num_boost_rounds=200,
+        metrics="auc",
+        n_split=4,
+        shuffle=True,
+        early_stop=20,
+        verbose=False,
+        **kwargs
+    ):
+
+        super(XGBoostClassifierHyperOpt, self).__init__(num_boost_rounds, **kwargs)
+
+        if isinstance(params, dict):
+            self.params = params
+        else:
+            raise TypeError("The input params must be a dictionary")
+
+        if isinstance(metrics, str):
+            self.metrics = metrics
+        else:
+            raise TypeError("The input metrics must be a str")
+
+        if isinstance(num_boost_rounds, int):
+            self.num_boost_rounds = num_boost_rounds
+        else:
+            raise TypeError("The input num_boost_rouns must be a int")
+
+        if isinstance(n_split, int):
+            self.n_split = n_split
+        else:
+            raise TypeError("The input n_split must be a int")
+
+        if isinstance(shuffle, bool):
+            self.shuffle = shuffle
+        else:
+            raise TypeError("The input shuffle must be a boolean")
+
+        if isinstance(early_stop, int):
+            self.early_stop = early_stop
+        else:
+            raise TypeError("The input early_stop must be a int")
+
+        if isinstance(verbose, bool):
+            self.verbose = verbose
+        else:
+            raise TypeError("The input verbose must be a boolean")
+
+    def process(self, X_train, Y_train, func_name, space, trials, algo, max_evals):
```

Comment:
The same here. func_name, space, trials, algo, max_evals, these all should have initialized values. 

I checked the example. you have:
def get_xgb_params():
    """ Define Parameter Space"""
    params  = {'nthread':4,
              'booster':'gbtree',
              'tree_method':'hist',
              'objective':'binary:logistic',
              'max_depth':hp.choice('max_depth',range(5,100,5)),
              'learning_rate':hp.quniform('learning_rate',0.01,0.50,0.01),
              'gamma':hp.quniform('gamma',0,0.50,0.01),
              'min_child_weight':hp.quniform('min_child_weight',1,10,1),
              'subsample':hp.quniform('subsample',0.1,1,0.01),
              'colsample_bytree': hp.quniform('colsample_bytree',0.1,1.0,0.01),
              'gamma': hp.quniform('gamma',0.0,1.0,0.01),
              'reg_alpha':hp.quniform('reg_alpha',0.0,1.0,0.01),
              'reg_lambda':hp.quniform('reg_lambda',0.0,1.0,0.01)
              }
    return params

This should be initialized inside the function. with like more logical values. For instance max_depth > 7 like it never happens.